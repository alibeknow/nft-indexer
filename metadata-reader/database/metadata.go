package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
