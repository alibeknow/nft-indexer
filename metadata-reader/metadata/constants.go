package metadata

type Chain string

const (
	ETH     Chain = "eth"
	POLYGON Chain = "polygon"
)

func GetMappings() map[string]interface{} {
	properties := make(map[string]interface{})

	properties["name"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["description"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["attributes"] = map[string]interface{}{
		"type":    "object",
		"enabled": true,
		"dynamic": true,
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type": "text",
			},
			"value": map[string]interface{}{
				"type": "text",
			},
		},
	}

	properties["chain"] = map[string]interface{}{
		"type":  "keyword",
		"index": true,
	}

	properties["contractAddress"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["metadata"] = map[string]interface{}{
		"type":    "object",
		"enabled": false,
		"dynamic": true,
	}

	return map[string]interface{}{
		"properties": properties,
	}
}
