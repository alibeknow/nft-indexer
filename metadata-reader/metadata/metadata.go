package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
