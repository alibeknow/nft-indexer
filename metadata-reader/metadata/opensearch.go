package metadata

import (
	"encoding/json"
	"fmt"
)

type MetadataDocument struct {
	Name            string                 `json:"name"`
	Description     string                 `json:"description"`
	ContractAddress string                 `json:"contractAddress"`
	Attributes      []map[string]string    `json:"attributes"`
	Chain           Chain                  `json:"chain"`
	Metadata        map[string]interface{} `json:"metadata"`
}

func MapToMetadataDocument(contractAddress string, chain Chain, rawMetadata string) (MetadataDocument, error) {
	var metadata map[string]interface{}
	var props struct {
		Attributes []map[string]interface{} `json:"attributes"`
		Properties map[string]interface{}   `json:"properties"`
	}

	err := json.Unmarshal([]byte(rawMetadata), &metadata)
	if err != nil {
		return MetadataDocument{}, err
	}

	_ = json.Unmarshal([]byte(rawMetadata), &props)

	attributes := []map[string]string{}
	for _, attribute := range props.Attributes {
		var key interface{}
		if value, found := attribute["trait_type"]; found {
			key = value
		}

		attributes = append(attributes, map[string]string{
			"key":   fmt.Sprint(key),
			"value": fmt.Sprint(attribute["value"]),
		})
	}

	for key, value := range props.Properties {
		b, err := json.Marshal(value)
		if err != nil {
			return MetadataDocument{}, err
		}

		attributes = append(attributes, map[string]string{
			"key":   key,
			"value": string(b),
		})
	}

	name, _ := metadata["name"].(string)
	description, _ := metadata["description"].(string)

	return MetadataDocument{
		Name:            name,
		Description:     description,
		ContractAddress: contractAddress,
		Attributes:      attributes,
		Chain:           chain,
		Metadata:        metadata,
	}, nil
}
