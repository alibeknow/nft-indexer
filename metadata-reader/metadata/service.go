package metadata

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"go.uber.org/zap"

	"metadata-reader/private/httpreader"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
)

var (
	metadataCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "metadata_counter",
			Help: "Number of received metadata",
		},
	)
	errCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "err_counter",
			Help: "Received errors count",
		},
	)
	tooManyReqCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "too_many_request_counter",
			Help: "Number of rate limit requests",
		},
	)
	failedReqCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "failed_request_counter",
			Help: "Number of failed requests",
		},
	)
	requestCounter = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "request_counter",
			Help: "Number of requests",
		},
	)
	storeMetadataDurationGauge = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "store_metadata_gauge",
			Help: "Store metadata duration",
		},
	)
)

func init() {
	prometheus.MustRegister(metadataCounter)
	prometheus.MustRegister(errCounter)
	prometheus.MustRegister(tooManyReqCounter)
	prometheus.MustRegister(failedReqCounter)
	prometheus.MustRegister(requestCounter)
	prometheus.MustRegister(storeMetadataDurationGauge)
}

type Config struct {
	Proxy   string
	Gateway string
	Retries int
	Timeout int
	Index   string
}

type Service struct {
	log        *zap.Logger
	db         DB
	httpReader *httpreader.HttpReader
	gateway    string
	opensearch *opensearch.OpenSearch
	indexName  string
}

func NewService(log *zap.Logger, config Config, db DB, search *opensearch.OpenSearch) (*Service, error) {
	var proxy *url.URL
	var err error

	if config.Proxy != "" {
		proxy, err = url.Parse(config.Proxy)
		if err != nil {
			return nil, err
		}
	}

	httpReader := httpreader.NewHttpReader(proxy, config.Retries, time.Duration(config.Timeout)*time.Second)

	return &Service{
		log:        log,
		db:         db,
		httpReader: httpReader,
		gateway:    config.Gateway,
		opensearch: search,
		indexName:  config.Index,
	}, nil
}

func (s *Service) Read(ctx context.Context, msg queue.MetadataMessage) (bool, error) {
	protocol := strings.SplitN(msg.TokenURI, ":", 2)[0]
	tokenRef := fmt.Sprintf("%s:%s:%s", msg.Blockchain, msg.ContractAddress, msg.TokenID)

	if protocol == "data" {
		tokenData := strings.SplitN(msg.TokenURI, ",", 2)
		if len(tokenData) < 2 {
			s.log.Debug("Inappropriate on-chain metadata format", zap.String("token", tokenRef))
			return false, nil
		}

		headWithoutDataPrefix := strings.Replace(tokenData[0], "data:", "", 1)

		dataTypeAndEncoding := strings.SplitN(headWithoutDataPrefix, ";", 2)
		if len(dataTypeAndEncoding) < 2 {
			s.log.Debug("Inappropriate on-chain metadata format", zap.String("token", tokenRef))
			return false, nil
		}

		if dataTypeAndEncoding[1] != "base64" {
			s.log.Debug("Inappropriate on-chain metadata encoding", zap.String("token", tokenRef))
			return false, nil
		}

		data, err := base64.StdEncoding.DecodeString(tokenData[1])
		if err != nil {
			s.log.Debug("Error decoding metadata", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}
		metadataCounter.Inc()
		startStoreToDB := time.Now()
		err = s.db.InsertOrUpdate(ctx, Metadata{
			ID:       tokenRef,
			Metadata: string(data),
			Type:     protocol,
		})
		storeMetadataDuration := time.Since(startStoreToDB).Microseconds()
		storeMetadataDurationGauge.Set(float64(storeMetadataDuration))
		if err != nil {
			s.log.Error("Error writing metadata to DB", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		if s.opensearch == nil {
			return false, nil
		}

		metadataDocument, err := MapToMetadataDocument(msg.ContractAddress, Chain(msg.Blockchain), string(data))
		if err != nil {
			s.log.Error("Error mapping metadata to opensearch document", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		b, err := json.Marshal(metadataDocument)
		if err != nil {
			s.log.Error("Error marshaling opensearch metadata document", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		_, err = s.opensearch.UpsertToIndex(ctx, s.indexName, tokenRef, strings.NewReader(string(b)))
		if err != nil {
			s.log.Error("Error writing metadata to opensearch index", zap.String("token", tokenRef), zap.Error(err))
			return false, err
		}

		s.log.Debug("Token is added to index", zap.String("token", tokenRef))

		return false, nil
	}

	uri := msg.TokenURI

	switch protocol {
	case "ipfs":
		trimmed := strings.TrimPrefix(uri, "ipfs://")
		trimmed = strings.TrimPrefix(trimmed, "ipfs/")
		uri = s.gateway + trimmed
	case "https":
		if strings.HasPrefix(msg.TokenURI, "https://ipfs.io/ipfs/") {
			trimmed := strings.TrimPrefix(msg.TokenURI, "https://ipfs.io/ipfs/")
			uri = s.gateway + trimmed
			break
		}
		if strings.HasPrefix(msg.TokenURI, "https://gateway.pinata.cloud/ipfs/") {
			trimmed := strings.TrimPrefix(msg.TokenURI, "https://gateway.pinata.cloud/ipfs/")
			uri = s.gateway + trimmed
			break
		}
		if strings.HasPrefix(msg.TokenURI, "https://api.opensea.io/api/") {
			replacer := strings.NewReplacer("{id}", strings.Replace(msg.TokenID, "0x", "", -1))
			uri = fmt.Sprintf(replacer.Replace(msg.TokenURI))
		}
	}

	resp, body, err := s.httpReader.Read(ctx, uri)
	requestCounter.Inc()
	if err != nil {
		s.log.Debug("HttpReader error", zap.String("token", tokenRef), zap.String("uri", uri), zap.String("originalUri", msg.TokenURI), zap.Error(err))
		errCounter.Inc()
		return false, err
	}

	if resp.StatusCode != http.StatusOK {
		s.log.Debug("Request failed with status code", zap.Int("statusCode", resp.StatusCode), zap.String("token", tokenRef), zap.String("uri", uri), zap.String("originalUri", msg.TokenURI))
		failedReqCounter.Inc()
		if resp.StatusCode == http.StatusTooManyRequests {
			tooManyReqCounter.Inc()
			return true, nil
		}
		return false, nil
	}

	metadataCounter.Inc()
	startStoreToDB := time.Now()
	err = s.db.InsertOrUpdate(ctx, Metadata{
		ID:       tokenRef,
		Metadata: string(body),
		Type:     protocol,
	})
	storeMetadataDuration := time.Since(startStoreToDB).Microseconds()
	storeMetadataDurationGauge.Set(float64(storeMetadataDuration))
	if err != nil {
		s.log.Error("Error writing metadata to DB", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	if s.opensearch == nil {
		return false, nil
	}

	metadataDocument, err := MapToMetadataDocument(msg.ContractAddress, Chain(msg.Blockchain), string(body))
	if err != nil {
		s.log.Error("Error mapping metadata to opensearch document", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	b, err := json.Marshal(metadataDocument)
	if err != nil {
		s.log.Error("Error marshaling opensearch metadata document", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	_, err = s.opensearch.UpsertToIndex(ctx, s.indexName, tokenRef, strings.NewReader(string(b)))
	if err != nil {
		s.log.Error("Error writing metadata to opensearch index", zap.String("token", tokenRef), zap.Error(err))
		return false, err
	}

	s.log.Debug("Token is added to index", zap.String("token", tokenRef))

	return false, nil
}
