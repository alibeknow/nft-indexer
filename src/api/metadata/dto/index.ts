export * from './nft-metadata-param.dto';
export * from './nfts-metadata-param.dto';
export * from './nfts-metadata-query.dto';
