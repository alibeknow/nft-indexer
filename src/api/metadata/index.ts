export * from './metadata.controller';
export * from './metadata.module';
export * from './metadata.service';
