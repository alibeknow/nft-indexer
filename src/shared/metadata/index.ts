export * from './metadata-repo-db';
export * from './metadata-repo-storage';
export * from './on-chain-reader';
export * from './ipfs-reader';
export * from './http-reader';
export * from './interfaces';
export * from './constants';
export * from './read-content-headers';
export * from './pinata-reader';
export * from './metadata-search';
