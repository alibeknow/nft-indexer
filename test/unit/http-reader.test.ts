import { HttpReader } from '@shared/metadata';
import axios from 'axios';
import { axiosGetResponse } from './fixtures/axios-get-reponse';

jest.setTimeout(40000);
jest.mock('axios');


class RateLimitError extends Error {
  public response = { status: 429 };
}

describe('HttpReader', () => {

  describe('[read] method', () => {
    it('should not apply retries if retries argument is 0', async () => {
      jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve(axiosGetResponse));

      const httpReader = new HttpReader();
      const tokenUri = 'https://ipfs.io/ipfs/QmapFPdNCAHMSKKcCLAfq3j9VGXJbvewwaC65tTkHrrL4T/265.json';
      const result = await httpReader.read(tokenUri);

      expect(axios.get).toHaveBeenNthCalledWith(1, tokenUri, { proxy: false, transformResponse: [], timeout: 0 });
      expect(result).toMatchSnapshot();
    });

    it('should not apply retries if retries argument is 0 and an error happened', async () => {
      const fakeErrorMsg = 'Some unexpected error...';

      jest.spyOn(axios, 'get').mockImplementation(() => {
        return new Promise((resolve, reject) => {
          setImmediate(() => {
            reject(new Error(fakeErrorMsg));
          });
        });
      });

      const httpReader = new HttpReader();
      const tokenUri = 'https://ipfs.io/ipfs/QmapFPdNCAHMSKKcCLAfq3j9VGXJbvewwaC65tTkHrrL4T/265.json';

      await expect(async () => {
        await httpReader.read(tokenUri);
      }).rejects.toThrow(fakeErrorMsg);
      expect(axios.get).toHaveBeenNthCalledWith(1, tokenUri, { proxy: false, transformResponse: [], timeout: 0 });
    });

    it('should apply retries in case of errors and if retries argument is > 0', async () => {
      let counter = 0;
      jest.spyOn(axios, 'get').mockImplementation(() => {
        return new Promise((resolve, reject) => {
          setImmediate(() => {
            if (counter < 3) {
              return reject(new RateLimitError('Rate limit error...'));
            } else {
              return resolve(axiosGetResponse);
            }
          });
          counter++;
        });
      });

      const httpReader = new HttpReader(3);
      const tokenUri = 'https://ipfs.io/ipfs/QmapFPdNCAHMSKKcCLAfq3j9VGXJbvewwaC65tTkHrrL4T/265.json';
      const result = await httpReader.read(tokenUri);

      expect(axios.get).toHaveBeenNthCalledWith(1, tokenUri, { proxy: false, transformResponse: [], timeout: 0 });
      expect(result).toMatchSnapshot();
    });
  });
});
