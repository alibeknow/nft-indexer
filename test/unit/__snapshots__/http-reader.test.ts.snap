// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpReader [read] method should apply retries in case of errors and if retries argument is > 0 1`] = `
{
  "attributes": [
    {
      "trait_type": "3D Background",
      "value": "Aquamarine",
    },
    {
      "trait_type": "Clothes",
      "value": "Green Christmas Sweater",
    },
    {
      "trait_type": "Earring",
      "value": "None",
    },
    {
      "trait_type": "Eyes",
      "value": "Open",
    },
    {
      "trait_type": "Hat",
      "value": "Bull Horn",
    },
    {
      "trait_type": "Mouth",
      "value": "Angry",
    },
  ],
  "image": "ipfs://QmXV7uB2Vo2wyrvTvhViRfzPiUopaw8fyrxDvFjQf7GTqe/265.png",
  "name": "Happy Koalas 265 - Test",
  "tokenId": 265,
}
`;

exports[`HttpReader [read] method should not apply retries if retries argument is 0 1`] = `
{
  "attributes": [
    {
      "trait_type": "3D Background",
      "value": "Aquamarine",
    },
    {
      "trait_type": "Clothes",
      "value": "Green Christmas Sweater",
    },
    {
      "trait_type": "Earring",
      "value": "None",
    },
    {
      "trait_type": "Eyes",
      "value": "Open",
    },
    {
      "trait_type": "Hat",
      "value": "Bull Horn",
    },
    {
      "trait_type": "Mouth",
      "value": "Angry",
    },
  ],
  "image": "ipfs://QmXV7uB2Vo2wyrvTvhViRfzPiUopaw8fyrxDvFjQf7GTqe/265.png",
  "name": "Happy Koalas 265 - Test",
  "tokenId": 265,
}
`;
